debian/dbinfo.1
.\" Automatically generated by Pandoc 1.19.2
.\"
.TH "DBINFO" "1" "" "MANUAL" ""
.hy
.SH NAME
.PP
dbinfo \- GBV Datenbankverzeichnis
.SH SYNOPSIS
.PP
The application is automatically started as service, listening on port
6006.
.IP
.nf
\f[C]
sudo\ service\ dbinfo\ {status|start|stop|restart}
\f[]
.fi
.SH INSTALLATION
.PP
The software is released as Debian package for Ubuntu 14.04 LTS.
Other Debian based distributions \f[I]might\f[] work too.
Releases can be found at <https://github.com/gbv/dbinfo/releases>
.PP
To install required dependencies either use a package manager such as
\f[C]gdebi\f[], manually install dependencies (inspectable via
\f[C]dpkg\ \-I\ dbinfo_*.deb\f[]):
.IP
.nf
\f[C]
sudo\ dpkg\ \-i\ ...\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ install\ dependencies
sudo\ dpkg\ \-i\ dbinfo_X.Y.Z_amd64.deb\ \ \ \ #\ change\ X.Y.Z
\f[]
.fi
.PP
After installation the service is available at localhost on port 6006.
Better put the service behind a reverse proxy to enable SSL and nice
URLs! In the following a reverse proxy mapping
<http://uri.gbv.de/database/> to <http://localhost:6006/> is assumed.
.SH USAGE
.PP
GBV Datenbankverzeichnis provides an overview of databases used by or
provided at VZG.
Information about databases can be retrieved in RDF, JSON, and CSV.
.SS Database organization
.PP
Each database is identified by an URI with prefix
<http://uri.gbv.de/database/> followed by a database key
(\f[B]dbkey\f[]).
Each dbkey consists of lowercase letters (\f[C]a\-z\f[]), digits
(\f[C]0\-9\f[]), and/or minus (\f[C]\-\f[]).
The part before the first minus, if given, identifies a database group.
For instance:
.IP \[bu] 2
<http://uri.gbv.de/database/vd17>: dbkey \f[C]vd17\f[]
.IP \[bu] 2
<http://uri.gbv.de/database/opac>: database group \f[C]opac\f[]
.IP \[bu] 2
<http://uri.gbv.de/database/opac-de-hil2>: database
\f[C]opac\-de\-hil2\f[] in group \f[C]opac\f[]
.PP
Information about databases groups is encoded in SKOS and can be
retrieved in JSKOS (\f[C]format=jsold\f[]).
A list of all database groups can be retrieved at the base URL:
.IP \[bu] 2
<http://uri.gbv.de/database/?format=jsonld>
.IP \[bu] 2
<http://uri.gbv.de/database/opac?format=jsonld>
.SS Database metadata
.PP
Information about databases can be retrieved in several RDF
serialization forms selecetd by HTTP content negotiation or via query
parameter \f[C]format\f[] (set to \f[C]ttl\f[], \f[C]json\f[],
\f[C]rdfxml\f[], \f[C]jsonld\f[] for RDF, or \f[C]html\f[] for HTML as
default).
The RDF vocabularies are documented at a page accessible via
<http://uri.gbv.de/database/ontology.html>.
A convenient data format for reuse outside of RDF applications is
JSON\-LD, specified at <http://uri.gbv.de/database/dbinfo.jsonld>:
.IP
.nf
\f[C]
curl\ http://uri.gbv.de/database/amb?format=jsonld
\f[]
.fi
.IP
.nf
\f[C]
{
\ \ \ "\@context"\ :\ "http://uri.gbv.de/database/dbinfo.jsonld",
\ \ \ "uri"\ :\ "http://uri.gbv.de/database/amb",
\ \ \ "title"\ :\ {
\ \ \ \ \ \ "de"\ :\ "Katalog\ der\ meereswissenschaftlichen\ Bibliotheken\ Deutschlands",
\ \ \ \ \ \ "en"\ :\ "German\ Association\ of\ Marine\ Science\ Libraries\ and\ Information\ Centers\ Catalogue"
\ \ \ },
\ \ \ "dbkey"\ :\ "amb",
\ \ \ ...
\f[]
.fi
.SS Database statistics
.PP
ppend \f[C]\&.csv\f[] to the Database URI to get statistics in CSV
format.
.SS Suggest service
.PP
The service further provides an OpenSearch Suggest AI at
\f[C]/api/dbkey\f[] to query database prefixes:
.IP
.nf
\f[C]
curl\ http://uri.gbv.de/database/api/dbkey?id=opac\-de\-91
\f[]
.fi
.IP
.nf
\f[C]
[\ "opac\-de\-91",
\ \ ["opac\-de\-916"],
\ \ ["Online\-Katalog\ der\ Ostfalia\ Hochschule\ für\ angewandte\ Wissenschaften"],
\ \ ["http://uri.gbv.de/database/opac\-de\-916"]\ ]
\f[]
.fi
.SH ADMINISTRATION
.SS Configuration
.PP
Config file \f[C]/etc/default/dbinfo\f[] only contains basic server
configuration in form of simple key\-values pairs:
.IP \[bu] 2
\f[C]PORT\f[] \- port number (required, 6006 by default)
.IP \[bu] 2
\f[C]WORKERS\f[] \- number of parallel connections (required, 5 by
default).
.PP
Additional configuration is placed in \f[C]/etc/dbinfo/config.yml\f[]
with the following fields:
.IP \[bu] 2
\f[C]unapi\f[] \- base URL of unAPI config server to get databases from
.IP \[bu] 2
\f[C]proxy\f[] \- optional list of IPs or IP ranges the service can run
behind (for logging the proxied request IPs instead of proxy IP).
.IP \[bu] 2
\f[C]base\f[] \- base URI (\f[C]http://uri.gbv.de/database/\f[] by
default)
.IP \[bu] 2
\f[C]stats\f[] \- directory to store statistics in
(\f[C]/etc/dbinfo/stats\f[] by default).
.PP
Restart is required after changes.
.PP
The script \f[C]/etc/dbinfo/diagram.r\f[] is used to generate diagrams.
It can be adjusted locally, so updates to this script must be applied
manually!
.SS Database Statistics
.PP
Database statistics are stored in \f[C]/etc/dbinfo/stats\f[] (or in
another directory specified in configuration) in a CSV file for each
database.
This directory should be backed up to retain history data.
Statistics is updated daily via cronjob in
\f[C]/etc/cron.daily/dbinfo\f[].
.SS Logging
.PP
Log files are located at \f[C]/var/log/dbinfo/\f[]:
.IP \[bu] 2
\f[C]error.log\f[]
.IP \[bu] 2
\f[C]access.log\f[]
.IP \[bu] 2
\f[C]mkstat.log\f[]
.SH CHANGES
.PP
See \f[C]debian/changelog\f[].
.SH SEE ALSO
.PP
The source code of dbinfo is managed in a public git repository at
<https://github.com/gbv/dbinfo>.
Please report bugs and feature request at
<https://github.com/gbv/dbinfo/issues>!
.PP
The Changelog is located in file \f[C]debian/changelog\f[].
.PP
Development guidelines are given in file \f[C]CONTRIBUTING.md\f[].


debian/changelog
dbinfo (0.9.6) unstable; urgency=medium
  * fully migrate from gbv.de to k10plus.de

 -- Jakob Voß <jakob.voss@gbv.de>  Wed, 16 Aug 2023 15:48:24 +0200

dbinfo (0.9.5) unstable; urgency=medium
  * fix IKT-Link with /

 -- Jakob Voß <jakob.voss@gbv.de>  Mon, 10 Jan 2022 09:28:24 +0200

dbinfo (0.9.4) unstable; urgency=medium

  * sort databases by key
  * fix kxp URLs (again)
  * replace AngularJS with jQuery

 -- Jakob Voß <jakob.voss@gbv.de>  Tue, 2 Jul 2019 13:25:25 +0200

dbinfo (0.9.3) unstable; urgency=medium

  * extend information about APIs
  * remove information about open/closed access

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 6 Jun 2019 09:08:25 +0200

dbinfo (0.9.2) unstable; urgency=medium

  * adjust SRU and IKT URLs for kxp databases

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 16 May 2019 12:20:15 +0200

dbinfo (0.9.1) unstable; urgency=medium

  * fix broken field host in kxp databases

 -- Jakob Voß <jakob.voss@gbv.de>  Mon, 29 Apr 2019 10:02:00 +0100

dbinfo (0.9.0) unstable; urgency=medium

  * add special handling of K10plus databases

 -- Jakob Voß <jakob.voss@gbv.de>  Wed, 24 Apr 2019 08:45:21 +0100

dbinfo (0.8.2) unstable; urgency=medium

  * fixed Debian init script
  * add JSKOS output (JSON-LD) for database groups 

 -- Jakob Voß <jakob.voss@gbv.de>  Tue, 10 Nov 2015 16:42:59 +0100

dbinfo (0.8.1) unstable; urgency=medium

  * show statistics only if available 
  * show version and link to GitHub
  * include CSV header (requires to manually update `etc/diagram.r`)

 -- Jakob Voß <jakob.voss@gbv.de>  Mon, 09 Nov 2015 10:01:39 +0100

dbinfo (0.8.0) unstable; urgency=medium

  * include creation of diagrams. Update requires manually
    copying `etc/diagram.r` to `/etc/dbinfo`

 -- Jakob Voß <jakob.voss@gbv.de>  Tue, 03 Nov 2015 14:25:52 +0100

dbinfo (0.7.1) unstable; urgency=medium

  * new JSON-LD based RDF serialization
  * add English titles
  * include public homepage (url field)
  * extended documentation

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 29 Oct 2015 11:30:00 +0000

dbinfo (0.7.0) unstable; urgency=medium

  * massive refactoring
  * additional suggest endpoint at /api/dbkey

 -- Jakob Voß <jakob.voss@gbv.de>  Tue, 22 Sep 2015 14:04:57 +0200

dbinfo (0.6.0-pre) stable; urgency=low

  * migrate to easier Debian packaging

 -- Jakob Voss <voss@gbv.de>  Mon, 26 Jan 2015 15:23:00 +0100


lib/App/DBInfo.pm
package App::DBInfo;
use v5.14;

use Log::Contextual::Easy::Default;

# Required TemplateToolkit plugins
use Template::Plugin::Number::Format;
use Template::Plugin::JSON::Escape;

use Plack::Middleware::TemplateToolkit qw(0.21);
use Plack::Middleware::Cached;
use RDF::Lazy qw(0.061);
use Plack::Middleware::Rewrite;
use RDF::Flow qw(0.175 :all);
use Plack::Builder;
use Plack::Request;
use URI::Escape;
use CHI;

use RDF::Trine qw(iri);
use RDF::Trine::Model;
use RDF::Trine::Parser;

use RDF::NS;
use constant NS => RDF::NS->new('20130930');

use App::DBInfo::Source;

use Plack::Builder;

use CHI;
use Plack::Middleware::Negotiate;
use Encode qw(encode);

use List::Util qw(first);
use YAML       ();

our $VERSION="0.9.6";

my $NEGOTIATE = Plack::Middleware::Negotiate->new(
    parameter => 'format',
    extension => 1,
    formats   => {
        nt     => { type    => 'text/plain' },
        rdf    => { type    => 'application/rdf+xml' },
        xml    => { type    => 'application/rdf+xml' },
        rdfxml => { type    => 'application/rdf+xml' },
        ttl    => { type    => 'text/turtle' },
        json   => { type    => 'application/rdf+json' },
        dbinfo => { type    => 'application/ld+json' },
        jsonld => { type    => 'application/ld+json' },
        html   => { type    => 'text/html' },
        _      => { charset => 'utf-8', }
    }
);

sub new {
    my $self = bless {}, shift;

    # load config file and set default values
    $self->{etcdir} = first { -e "$_/config.yml" } '/etc/dbinfo', 'etc';
    $self->{config} = YAML::LoadFile( $self->{etcdir} . "/config.yml" );

    $self->{config}{base}  //= 'http://uri.gbv.de/database/';
    $self->{config}{stats} //= $self->{etcdir} . "/stats";

    die "Missing unapi URL in config file\n" unless $self->{config}{unapi};
    $self->{config}{unapi} =~ s{/$}{};

    # TODO: get rid of RDF::Flow, include App::DBInfo::Source here
    $self->{source} = App::DBInfo::Source->new;

    # TODO: avoid hack
    $self->{source}{config} = $self->{config};
    $self->{source}->load;

    $self->{rdfsource} =
      RDF::Flow::Cached->new( $self->{source},
        CHI->new( driver => 'Memory', global => 1, expires_in => '1 hour' ) );

    $self->{app} = builder {

        enable_if { $self->{config}{proxy} } 'Plack::Middleware::XForwardedFor',
          trust => $self->{config}{proxy};

        enable 'Static',
          root         => $self->{config}{stats},
          path         => qr{\.(png|csv)$},
          pass_through => 1;

        enable 'Static',
          root         => 'public',
          path         => qr{\.(css|png|gif|js|ico|jsonld)$},
          pass_through => 0;

        # cache everything else for 10 seconds. TODO: set cache time
        enable 'Cached',
          cache => CHI->new(
            driver     => 'Memory',
            global     => 1,
            expires_in => '10 seconds'
          );

        enable 'JSONP';

        mount '/api/dbkey' => sub {
            my $id     = Plack::Request->new( $_[0] )->param('id') // '';
            my $result = $self->{source}->suggest_dbkey($id);
            my $json   = JSON->new->encode($result);
            return [ 200, [ "Content-Type" => "text/javascript" ], [$json] ];
        };

        mount '/' => $NEGOTIATE->wrap( sub { $self->core(@_) } );
    };

    return $self;
}

sub to_app {
    my $self = shift;
    return sub { $self->{app}->(@_) }
}

sub core {
    my ( $self, $env ) = @_;
    my $req = Plack::Request->new($env);

    # construct request URI
    my $base = $self->{config}{base} || $req->base;

    my $path = $req->path;
    $path =~ s/^\///;
    my $uri = $base . $path;

    # retrieve RDF
    $env->{'rdflow.uri'} = $uri;
    my $rdf = $self->{rdfsource}->retrieve($env);

    my $format = $env->{'negotiate.format'} // 'html';

    # serialize and return RDF data
    unless ( grep { $format eq $_ } qw(html dbinfo jsonld) ) {
        if ( $env->{'rdflow.error'} ) {
            return [
                500,
                [ 'Content-Type' => 'text/plain' ],
                [ $env->{'rdflow.error'} ]
            ];
        }

        my $ser = {
            nt     => 'NTriples',
            rdf    => 'RDFXML',
            xml    => 'RDFXML',
            rdfxml => 'RDFXML',
            ttl    => 'Turtle',
            json   => 'RDFJSON',
        };
        my $serializer = RDF::Trine::Serializer->new( $ser->{$format} );

        my $rdf_data;

        if ( UNIVERSAL::isa( $rdf, 'RDF::Trine::Model' ) ) {
            $rdf_data = $serializer->serialize_model_to_string($rdf);
        }
        elsif ( UNIVERSAL::isa( $rdf, 'RDF::Trine::Iterator' ) ) {
            $rdf_data = $serializer->serialize_iterator_to_string($rdf);
        }
        if ($rdf_data) {
            $rdf_data = encode( 'utf8', $rdf_data );
            my $headers = [];
            $NEGOTIATE->add_headers( $headers, $format );
            return [ 200, $headers, [$rdf_data] ];
        }
    }

    # initialize HTML format via Template
    my $vars = $env->{'tt.vars'} || {};

    $vars->{'formats'} = [qw(ttl json rdfxml jsonld)];

    my $lazy = RDF::Lazy->new( $rdf, namespaces => NS );

    if ( $rdf and $rdf->size ) {
        if ( $format =~ /^(dbinfo|jsonld)$/ ) {
            $vars->{'JSON_TRUE'}  = JSON::true;
            $vars->{'JSON_FALSE'} = JSON::false;

            $env->{'tt.path'} = '/jsonld.json';

        }
        elsif ( $uri eq $self->{config}{base} ) {

            # ...
        }
        elsif ( $lazy->resource($uri)->type('skos:Concept') ) {

            # show database group/prefix
            $env->{'tt.path'} = '/prefix.html';
        }
        else {
            # show database
            $env->{'tt.path'} = '/database.html';
        }
    }

    $vars->{uri} = $lazy->resource($uri);

    $vars->{error}     = $env->{'rdflow.error'};
    $vars->{timestamp} = $env->{'rdflow.timestamp'};
    $vars->{cached}    = 1 if $env->{'rdflow.cached'};
    $vars->{version}   = $VERSION;
    $vars->{config}    = $self->{config};

    $env->{'tt.vars'} = $vars;

    state $app = Plack::Middleware::TemplateToolkit->new(
        INCLUDE_PATH => 'public',
        RELATIVE     => 1,
        INTERPOLATE  => 1,
        pass_through => 0,
        vars         => { base => '.' },    # request.base.remove('/$')
        404          => '404.html',
        500          => '500.html'
    )->to_app;

    return $app->($env);
}

1;
